File: apps/web/lib/api/expressions.ts
import useSWR from "swr";

import { apiRequestAuth, fetcher } from "@/lib/api/fetcher";
import { GEOAPI_BASE_URL, PROCESSES_BASE_URL } from "@/lib/constants";

// Types
export interface ValidationError {
  message: string;
  code: string;
  position?: number | null;
  suggestion?: string | null;
}

export interface ValidateExpressionRequest {
  expression: string;
  column_names: string[];
  column_types?: Record<string, string> | null;
  geometry_column?: string | null;
}

export interface ValidateExpressionResponse {
  valid: boolean;
  expression: string;
  errors: ValidationError[];
  warnings: string[];
  referenced_columns: string[];
  used_functions: string[];
}

export interface PreviewRow {
  row_number: number;
  result: string | number | boolean | null;
  context: Record<string, unknown>;
}

export interface PreviewExpressionRequest {
  expression: string;
  where_clause?: string | null;
  limit?: number;
}

export interface PreviewExpressionResponse {
  success: boolean;
  expression: string;
  result_type?: string | null;
  rows: PreviewRow[];
  column_names: string[];
  error?: string | null;
}

export interface FunctionParameter {
  name: string;
  type: string;
  description_key: string;
  optional: boolean;
}

export interface FunctionDoc {
  name: string;
  category: string;
  syntax: string;
  example: string;
  description_key: string;
  parameters: FunctionParameter[];
}

export interface FunctionListResponse {
  functions: Record<string, FunctionDoc[]>;
  total: number;
}

// API functions
const EXPRESSIONS_API_URL = `${GEOAPI_BASE_URL}/expressions`;

/**
 * Validate an expression with explicit column names
 */
export async function validateExpression(
  request: ValidateExpressionRequest
): Promise<ValidateExpressionResponse> {
  const response = await apiRequestAuth(`${EXPRESSIONS_API_URL}/validate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to validate expression");
  }

  return response.json();
}

/**
 * Validate an expression for a specific collection (layer)
 */
export async function validateExpressionForCollection(
  collectionId: string,
  expression: string
): Promise<ValidateExpressionResponse> {
  const response = await apiRequestAuth(`${EXPRESSIONS_API_URL}/validate/${collectionId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ expression }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to validate expression");
  }

  return response.json();
}

/**
 * Preview an expression result for a collection
 */
export async function previewExpression(
  collectionId: string,
  request: PreviewExpressionRequest
): Promise<PreviewExpressionResponse> {
  const response = await apiRequestAuth(`${EXPRESSIONS_API_URL}/preview/${collectionId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Failed to preview expression");
  }

  return response.json();
}

/**
 * Hook to fetch all available functions
 */
export function useExpressionFunctions(enabled: boolean = true) {
  const { data, isLoading, error } = useSWR<FunctionListResponse>(
    enabled ? `${EXPRESSIONS_API_URL}/functions` : null,
    fetcher
  );

  return {
    functions: data?.functions,
    total: data?.total,
    isLoading,
    isError: error,
  };
}

/**
 * Hook to fetch functions by category
 */
export function useExpressionFunctionsByCategory(category: string) {
  const { data, isLoading, error } = useSWR<FunctionDoc[]>(
    category ? [`${EXPRESSIONS_API_URL}/functions/${category}`] : null,
    fetcher
  );

  return {
    functions: data,
    isLoading,
    isError: error,
  };
}

// API URL for processes (aggregation stats)
const PROCESSES_API_URL = `${PROCESSES_BASE_URL}/processes`;

export interface ExpressionPreviewResult {
  value: number | null;
  total_count: number;
  error?: string;
  items?: Array<{ grouped_value: string | number | null; operation_value: number }>;
}

/**
 * Preview an expression result using the aggregation-stats endpoint.
 * This shows how the expression will be computed when used in widgets.
 */
export async function previewExpressionAsAggregation(
  collectionId: string,
  expression: string,
  groupByColumn?: string | null,
  limit: number = 10
): Promise<ExpressionPreviewResult> {
  const inputs: Record<string, unknown> = {
    collection: collectionId,
    operation: "expression",
    operation_column: expression,
    limit: limit,
  };

  if (groupByColumn) {
    inputs.group_by_column = groupByColumn;
  }

  const response = await apiRequestAuth(`${PROCESSES_API_URL}/aggregation-stats/execution`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ inputs }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail?.detail || error.detail || "Failed to preview expression");
  }

  const result = await response.json();
  return {
    value: result.items?.[0]?.operation_value ?? null,
    total_count: result.total_count || 0,
    items: result.items || [],
  };
}

// Category display names and icons for UI
export const FUNCTION_CATEGORIES = {
  math: { labelKey: "category_math", icon: "calculator" },
  string: { labelKey: "category_string", icon: "text_fields" },
  date: { labelKey: "category_date", icon: "calendar_today" },
  aggregate: { labelKey: "category_aggregate", icon: "functions" },
  window: { labelKey: "category_window", icon: "view_column" },
  conditional: { labelKey: "category_conditional", icon: "call_split" },
} as const;

export type FunctionCategory = keyof typeof FUNCTION_CATEGORIES;

File: apps/geoapi/src/geoapi/routers/expressions.py
"""Expression builder router for validation and preview.

Provides endpoints for the Formula Builder feature:
- Validate expressions before applying them
- Preview expression results
- Get available functions by category
"""

import logging

from fastapi import APIRouter, HTTPException
from goatlib.utils.expressions import (
    FUNCTION_REGISTRY,
    ExpressionEvaluator,
    ExpressionValidator,
    FunctionCategory,
)
from pydantic import BaseModel, Field

from geoapi.dependencies import LayerInfoDep
from geoapi.ducklake_pool import ducklake_pool
from geoapi.services.layer_service import layer_service

router = APIRouter(prefix="/expressions", tags=["Expressions"])
logger = logging.getLogger(__name__)


# Request/Response models
class ValidateExpressionRequest(BaseModel):
    """Request to validate an expression."""

    expression: str = Field(..., description="The SQL expression to validate")
    column_names: list[str] = Field(
        ..., description="List of valid column names in the table"
    )
    column_types: dict[str, str] | None = Field(
        None,
        description="Dict mapping column names to their JSON types "
        "(e.g., {'population': 'integer', 'name': 'string'})",
    )
    geometry_column: str | None = Field(
        "geometry", description="Name of the geometry column"
    )


class ValidateExpressionResponse(BaseModel):
    """Response from expression validation."""

    valid: bool = Field(..., description="Whether the expression is valid")
    expression: str = Field(..., description="The validated expression")
    errors: list[dict] = Field(
        default_factory=list, description="List of validation errors"
    )
    warnings: list[str] = Field(
        default_factory=list, description="List of validation warnings"
    )
    referenced_columns: list[str] = Field(
        default_factory=list, description="Columns referenced in the expression"
    )
    used_functions: list[str] = Field(
        default_factory=list, description="Functions used in the expression"
    )


class PreviewExpressionRequest(BaseModel):
    """Request to preview an expression result."""

    expression: str = Field(..., description="The SQL expression to preview")
    where_clause: str | None = Field(
        None, description="Optional WHERE clause to filter rows"
    )
    limit: int = Field(10, ge=1, le=100, description="Number of rows to preview")


class PreviewRow(BaseModel):
    """A single row in the preview result."""

    row_number: int = Field(..., description="Row number")
    result: str | int | float | bool | None = Field(
        ..., description="Expression result"
    )
    context: dict = Field(default_factory=dict, description="Referenced column values")


class PreviewExpressionResponse(BaseModel):
    """Response from expression preview."""

    success: bool = Field(..., description="Whether the preview succeeded")
    expression: str = Field(..., description="The previewed expression")
    result_type: str | None = Field(None, description="DuckDB type of the result")
    rows: list[PreviewRow] = Field(default_factory=list, description="Preview rows")
    column_names: list[str] = Field(
        default_factory=list, description="Column names in context"
    )
    error: str | None = Field(None, description="Error message if preview failed")


class FunctionDocResponse(BaseModel):
    """Function documentation response."""

    name: str = Field(..., description="Function name")
    category: str = Field(..., description="Function category")
    syntax: str = Field(..., description="Function syntax")
    example: str = Field(..., description="Usage example")
    description_key: str = Field(..., description="i18n key for description")
    parameters: list[dict] = Field(
        default_factory=list, description="Parameter documentation"
    )


class FunctionListResponse(BaseModel):
    """Response with list of available functions."""

    functions: dict[str, list[FunctionDocResponse]] = Field(
        ..., description="Functions grouped by category"
    )
    total: int = Field(..., description="Total number of functions")


class ValidateExpressionSimpleRequest(BaseModel):
    """Simple request to validate an expression (for use with collection endpoint)."""

    expression: str = Field(..., description="The SQL expression to validate")


@router.post(
    "/validate",
    summary="Validate an expression",
    response_model=ValidateExpressionResponse,
)
async def validate_expression(
    request: ValidateExpressionRequest,
) -> ValidateExpressionResponse:
    """Validate a SQL expression against the function whitelist and column names.

    This endpoint checks:
    - Syntax correctness
    - Security (no DDL, DML, file access, etc.)
    - Function whitelist compliance
    - Column reference validity
    - Function parameter type compatibility (if column_types provided)

    Returns validation result with any errors and suggestions.
    """
    validator = ExpressionValidator(
        column_names=request.column_names,
        geometry_column=request.geometry_column,
        column_types=request.column_types,
    )
    result = validator.validate(request.expression)

    return ValidateExpressionResponse(
        valid=result.valid,
        expression=result.expression,
        errors=[
            {
                "message": e.message,
                "code": e.code,
                "position": e.position,
                "suggestion": e.suggestion,
            }
            for e in result.errors
        ],
        warnings=result.warnings,
        referenced_columns=list(result.referenced_columns),
        used_functions=list(result.used_functions),
    )


@router.post(
    "/validate/{collectionId}",
    summary="Validate an expression for a collection",
    response_model=ValidateExpressionResponse,
)
async def validate_expression_for_collection(
    layer_info: LayerInfoDep,
    request: ValidateExpressionSimpleRequest,
) -> ValidateExpressionResponse:
    """Validate a SQL expression for a specific collection.

    Automatically uses the collection's column names and types for validation,
    including function parameter type checking.
    """
    # Get layer metadata
    metadata = await layer_service.get_layer_metadata(layer_info)
    if not metadata:
        raise HTTPException(status_code=404, detail="Collection not found")

    validator = ExpressionValidator(
        column_names=metadata.column_names,
        geometry_column=metadata.geometry_column,
        column_types=metadata.column_types,
    )
    result = validator.validate(request.expression)

    return ValidateExpressionResponse(
        valid=result.valid,
        expression=result.expression,
        errors=[
            {
                "message": e.message,
                "code": e.code,
                "position": e.position,
                "suggestion": e.suggestion,
            }
            for e in result.errors
        ],
        warnings=result.warnings,
        referenced_columns=list(result.referenced_columns),
        used_functions=list(result.used_functions),
    )


@router.post(
    "/preview/{collectionId}",
    summary="Preview expression result",
    response_model=PreviewExpressionResponse,
)
async def preview_expression(
    layer_info: LayerInfoDep,
    request: PreviewExpressionRequest,
) -> PreviewExpressionResponse:
    """Preview an expression result on actual data.

    Runs the expression on a sample of rows and returns:
    - The computed result for each row
    - Referenced column values for context
    - The inferred result type

    The expression is validated before execution.
    """
    try:
        # Get layer metadata for validation
        metadata = await layer_service.get_layer_metadata(layer_info)
        if not metadata:
            raise HTTPException(status_code=404, detail="Collection not found")

        # Use the pool's connection context manager
        with ducklake_pool.connection() as con:
            evaluator = ExpressionEvaluator(
                con=con,
                table_name=layer_info.full_table_name,
                column_names=metadata.column_names,
                geometry_column=metadata.geometry_column,
            )
            result = evaluator.preview(
                expression=request.expression,
                where_clause=request.where_clause,
                limit=request.limit,
            )

        return PreviewExpressionResponse(
            success=result.success,
            expression=result.expression,
            result_type=result.result_type,
            rows=[
                PreviewRow(
                    row_number=row.row_number,
                    result=row.result,
                    context=row.context,
                )
                for row in result.rows
            ],
            column_names=result.column_names,
            error=result.error,
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.exception("Error previewing expression")
        return PreviewExpressionResponse(
            success=False,
            expression=request.expression,
            error=str(e),
        )


@router.get(
    "/functions",
    summary="Get available functions",
    response_model=FunctionListResponse,
)
async def get_functions() -> FunctionListResponse:
    """Get all available functions grouped by category.

    Returns the complete function registry with documentation
    for each function, including syntax, examples, and parameters.
    """
    functions_by_category: dict[str, list[FunctionDocResponse]] = {}

    for func_name, func_doc in FUNCTION_REGISTRY.items():
        category = func_doc.category.value

        if category not in functions_by_category:
            functions_by_category[category] = []

        functions_by_category[category].append(
            FunctionDocResponse(
                name=func_doc.name,
                category=category,
                syntax=func_doc.syntax,
                example=func_doc.example,
                description_key=func_doc.description_key,
                parameters=[
                    {
                        "name": p.name,
                        "type": p.type,
                        "description_key": p.description_key,
                        "optional": p.optional,
                    }
                    for p in func_doc.parameters
                ],
            )
        )

    # Sort functions within each category
    for category in functions_by_category:
        functions_by_category[category].sort(key=lambda f: f.name)

    return FunctionListResponse(
        functions=functions_by_category,
        total=len(FUNCTION_REGISTRY),
    )


@router.get(
    "/functions/{category}",
    summary="Get functions by category",
    response_model=list[FunctionDocResponse],
)
async def get_functions_by_category(
    category: FunctionCategory,
) -> list[FunctionDocResponse]:
    """Get functions for a specific category.

    Available categories:
    - math: Mathematical functions (ABS, ROUND, FLOOR, etc.)
    - string: String manipulation (UPPER, LOWER, TRIM, etc.)
    - date: Date/time functions (NOW, DATE_PART, etc.)
    - aggregate: Aggregation functions (SUM, AVG, COUNT, etc.)
    - window: Window functions (ROW_NUMBER, RANK, etc.)
    - spatial: Spatial functions (ST_Area, ST_Length, etc.)
    - conditional: Conditional logic (CASE, COALESCE, NULLIF, etc.)
    """
    result = []
    for func_name, func_doc in FUNCTION_REGISTRY.items():
        if func_doc.category == category:
            result.append(
                FunctionDocResponse(
                    name=func_doc.name,
                    category=func_doc.category.value,
                    syntax=func_doc.syntax,
                    example=func_doc.example,
                    description_key=func_doc.description_key,
                    parameters=[
                        {
                            "name": p.name,
                            "type": p.type,
                            "description_key": p.description_key,
                            "optional": p.optional,
                        }
                        for p in func_doc.parameters
                    ],
                )
            )

    result.sort(key=lambda f: f.name)
    return result
